service/availability_check.go

```diff
@@ -0,0 +1,154 @@
+package service
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/RedHatInsights/sources-api-go/config"
+	"github.com/RedHatInsights/sources-api-go/kafka"
+	l "github.com/RedHatInsights/sources-api-go/logger"
+	m "github.com/RedHatInsights/sources-api-go/model"
+	"github.com/RedHatInsights/sources-api-go/util"
+)
+
+type availabilityCheckRequester struct{}
+
+type availabilityChecker interface {
+	ApplicationAvailabilityCheck(source *m.Source)
+	EndpointAvailabilityCheck(source *m.Source)
+}
+
+var (
+	// storing the satellite topic here since it doesn't change after initial
+	// startup.
+	satelliteTopic = config.Get().KafkaTopic("platform.topological-inventory.operations-satellite")
+	// default availability checker instance
+	ac availabilityChecker = &availabilityCheckRequester{}
+)
+
+// requests both types of availability checks for a source
+func RequestAvailabilityCheck(source *m.Source) {
+	l.Log.Infof("Requesting Availability Check for Source [%v]", source.ID)
+
+	if len(source.Applications) != 0 {
+		ac.ApplicationAvailabilityCheck(source)
+	}
+
+	if len(source.Endpoints) != 0 {
+		ac.EndpointAvailabilityCheck(source)
+	}
+
+	l.Log.Infof("Finished Publishing Availability Messages for Source %v", source.ID)
+}
+
+// sends off an availability check http request for each of the source's
+// applications
+func (acr availabilityCheckRequester) ApplicationAvailabilityCheck(source *m.Source) {
+	for _, app := range source.Applications {
+		l.Log.Infof("Requesting Availability Check for Application %v", app.ID)
+
+		uri := app.ApplicationType.AvailabilityCheckURL()
+		if uri == nil {
+			l.Log.Warnf("Failed to fetch availability check url for [%v] - continuing", app.ApplicationType.Name)
+			continue
+		}
+
+		requestAvailabilityCheck(source, &app, uri)
+	}
+}
+
+func requestAvailabilityCheck(source *m.Source, app *m.Application, uri *url.URL) {
+	body := map[string]string{"source_id": strconv.FormatInt(app.SourceID, 10)}
+	raw, err := json.Marshal(body)
+	if err != nil {
+		l.Log.Warnf("Failed to marshal source body for [%v] - continuing", app.SourceID)
+		return
+	}
+
+	// spin up a 10 second context to limit the time spent waiting on a response
+	ctx, done := context.WithTimeout(context.Background(), 10*time.Second)
+	defer done()
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, uri.String(), bytes.NewBuffer(raw))
+	if err != nil {
+		l.Log.Warnf("Failed to make request for application [%v], uri [%v]", app.ID, uri.String())
+		return
+	}
+
+	req.Header.Add("x-rh-sources-account-number", source.Tenant.ExternalTenant)
```

Comment:
Should we add `x-rh-sources-account-number` header also to Kafka message ?

Reply:
I'm not sure if satellite uses it but it may - so I added it just in case.